x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -842,17 +848,47 @@ protected LogicalPlan rule(LogicalPlan p) {
         }
     }
 
-    /**
-     * A COUNT in a local relation will always be 1.
-     */
-    private static class ReplaceCountInLocalRelation extends OptimizerRule<Aggregate> {
+    static class ReplaceAggregationsInLocalRelations extends OptimizerRule<UnaryPlan> {
 
         @Override
-        protected LogicalPlan rule(Aggregate a) {
-            boolean hasLocalRelation = a.anyMatch(LocalRelation.class::isInstance);
+        protected LogicalPlan rule(UnaryPlan plan) {
+            if (plan instanceof Aggregate || plan instanceof Filter) {
+                // This optimization is only correct if `plan` has a LocalRelation descendant that is not filtered.
+                // Because PushDownAndCombineFilters might push down filters to become descendants of `plan`, this optimization
+                // must never run before PushDownAndCombineFilters.
+                LocalRelation relation = unfilteredLocalRelation(plan.child());
+                if (relation != null) {
+                    long count = (relation.executable() instanceof EmptyExecutable ? 0L : 1L);
+
+                    return plan.transformExpressionsDown(AggregateFunction.class, aggregateFunction -> {
+                        if (aggregateFunction instanceof Count) {
+                            return new Literal(Source.EMPTY, count, DataTypes.LONG);
+                        } else if (count == 0) {
+                            return new Literal(Source.EMPTY, null, aggregateFunction.dataType());
+                        } else {
+                            // note, most aggregation functions have already been substituted by something like
+                            // IIF(COUNT(1) == 0, literal, literal) in ReplaceAggregatesWithLiterals
+                            return aggregateFunction;
+                        }
+                    });
+                }
+            }
 
-            return hasLocalRelation ? a.transformExpressionsDown(Count.class, c -> new Literal(c.source(), 1, c.dataType())) : a;
+            return plan;
+        }
+
+        private LocalRelation unfilteredLocalRelation(LogicalPlan plan) {
```

Comment:
See the comment above - instead of checking for the presence of a filter, simply check there's always a `LocalRelation`. That is, the filters have to be folded/elided for this to work as should be the case of any other block.
Any cases that you are thinking of?

Reply:
For queries like `SELECT COUNT(*) WHERE FALSE HAVING COUNT(*) = 0` you need to make sure that you only apply the replacement of `COUNT(*)` on unfiltered relations:

```
Having[count(*) = 0]
\ Aggregate[count(*)]
 \ Filter[false]
  \ Local[Single]
```

`Filter` needs to be pruned first which does not always happen because the foldings generally happen before the prunings.